Core IPC messaging rewrite.

kinetiknz opened the pull request:
Replaces tokio 0.1 with mio 0.7-based system.  Full description in [old draft PR](https://github.com/mozilla/audioipc-2/pull/123).

Part 1 includes the new IPC core, intended to replace the existing tokio 0.1 IPC core.  Part 2 integrates the new IPC core into the rest of the server and client code.  Part 3 removes the old tokio 0.1 core (including resolving mio 0.6 vs 0.7 dependency duplication).

These changes will be initially merged into the https://github.com/mozilla/audioipc-2/tree/macos branch, which can then be integrated with Gecko for the macOS build only.  Once any remaining issues have been resolved, the macos branch will be merged to master and Linux and Windows will switch to the new IPC core.

ChunMinChang reviewed (commented):
I haven't read the whole code yet. I'll finish it next week. Just leaving some thoughts about the documents for now and some comments before I took the PTO last week.

ChunMinChang commented on audioipc/src/rpccore.rs:
```diff
@@ -0,0 +1,219 @@
+    let proxy = Proxy { handle: None, tx };
+    (proxy, rx)
+}
+
```
It would be great if we can have some comments stating the relationship between `ClientHandler` and `ServerHandler`:
```
ClientHandler       ServerHandler
    produce ----------> consume 
    consume <---------- produce
```

ChunMinChang commented on audioipc/src/ipccore.rs:
```diff
@@ -0,0 +1,717 @@
+        let handler = make_server::<S>(server);
+        let driver = Box::new(FramedDriver::new(handler));
+        self.add_connection(server_pipe, driver)?;
+        Ok(())
```
nit: I find using `Ok(self.add_connection(server_pipe, driver)?)` and adding a `trace!(..)` printing the token here helpful for me to understand the code better. But I don't have a strong opinion.

kinetiknz replied:
Done.

ChunMinChang reviewed (commented):
The change here is huge. Here are some initial thoughts after reading the code for now. I'll need to look deeper.

ChunMinChang commented on server/src/server.rs:
```diff
@@ -332,7 +321,7 @@ impl Drop for ServerStream {
 }
 
 struct CubebServerCallbacks {
```
This is unrelated to the changes here, but could we rename this to `DeviceCollectionChangeCallback` or `DeviceCallback`, or something similar? `callback` appears in lots of the classes or variables in this crate, and each of them has a different meaning. I find it hard to read know what this is without jumping into the definitions. This bothers me when reading the code here.

kinetiknz replied:
That's clearer, I've renamed it.

ChunMinChang commented on client/src/context.rs:
```diff
@@ -160,67 +143,44 @@ impl rpc::Server for DeviceCollectionServer {
 
-        let server_stream =
-            unsafe { audioipc::MessageStream::from_raw_handle(params.server_connection) };
+        let server_stream = unsafe { sys::Pipe::from_raw_handle(params.server_connection) };
```
I am not sure why this is named as `server_stream`. `ServerStream` is the cubeb stream on the ipc server-side/parent process. This should be the handle of `client_pipe` created in ipc server-side. Could we name this `server_connection`?

kinetiknz replied:
Done.

ChunMinChang reviewed (changes requested):
The readability is a lot better now, awesome! I enjoy reading the changes proposed here. I'm just holding this off until the questions here are answered. Some of the comments are unrelated to the changes made here but I think it's worth addressing those issues while we are trying to make the code easier to be maintained.

kinetiknz commented on ARCHITECTURE.md:
```diff
@@ -0,0 +1,31 @@
+- Remove existing tokio 0.1 code and deps (once all platforms switched over)
+- Delete tokio_named_pipes, tokio_uds_stream
+- Delete messagestream_*.rs
+- Delete AsyncSendMsg/AsyncRecvMsg traits/impls
```
Good idea.  I've removed the TODOs since they're complete (and it's better to track new ones in issues or as a TODO in the code for small items), and moved the contents of  ARCHITECTURE to README (with some rewording).

kinetiknz commented on ARCHITECTURE.md:
```diff
@@ -0,0 +1,31 @@
+        cubeb callbacks to communicate with client-side cubeb streams via EventLoop
+- Client side:
+    - EventLoop thread with a single connection, communicating with server
+    - RPC Proxy per cubeb stream, remoting cubeb API calls to the server (via EventLoop)
```
Rewritten per suggestions, hopefully it's somewhat clearer.  Happy to further add or clarify things in this PR as needed, but I'd also like continue expanding this document in a later PR along with reworking naming conventions of various parts of the code to reduce the confusion between the multiple uses of server, client, and other overloaded terms.

kinetiknz replied:
https://github.com/mozilla/audioipc-2/issues/133

kinetiknz commented on audioipc/src/ipccore.rs:
```diff
@@ -0,0 +1,575 @@
+    events: Events,
+    waker: Arc<Waker>,
+    connections: Slab<Connection>,
+    requests: mpsc::Receiver<Request>,
```
Renamed for clarity.

kinetiknz commented on audioipc/src/ipccore.rs:
```diff
@@ -0,0 +1,575 @@
+
+        // If the connection is readable, read into inbound and pass to driver for processing until all ready data
+        // has been consumed.
+        //let done = if let Some(event) = event {
```
Done.

kinetiknz commented on audioipc/src/ipccore.rs:
```diff
@@ -0,0 +1,717 @@
+    #[derive(Debug, Serialize, Deserialize, PartialEq)]
+    enum TestServerMessage {
+        TestRequest1,
+        TestRequest2,
```
Removed.

kinetiknz commented on audioipc/src/ipccore.rs:
```diff
@@ -0,0 +1,717 @@
+    #[derive(Debug, Serialize, Deserialize, PartialEq)]
+    enum TestClientMessage {
+        TestResponse1,
+        TestResponse2,
```
Removed.

kinetiknz commented on audioipc/src/ipccore.rs:
```diff
@@ -0,0 +1,717 @@
+        name: String,
+        stack_size: Option<usize>,
+        after_start: F1,
+        before_stop: F2,
```
All existing callers supply `F1` and `F2`.  Making them optional clutters the call site since you need to explicitly specify the type of `F1` and `F2` in the `None` case - that could be solved with a builder pattern, but since they're currently always specified it doesn't seem worth the added complexity.

kinetiknz commented on audioipc/src/ipccore.rs:
```diff
@@ -0,0 +1,717 @@
+            .expect("client EventLoopThread");
+        let client_handle = client_elt.handle();
+
+        let client_pipe = unsafe { sys::Pipe::from_raw_handle(client_pipe.into_raw()) };
```
Good idea, done.

kinetiknz commented on audioipc/src/rpccore.rs:
```diff
@@ -0,0 +1,219 @@
+    let (tx, rx) = mpsc::channel();
+    let proxy = Proxy { handle: None, tx };
+    (proxy, rx)
+}
```
That's tidier, thanks.

kinetiknz commented on audioipc/src/rpccore.rs:
```diff
@@ -0,0 +1,219 @@
+}
+
+pub fn make_client<C: Client>() -> (ClientHandler<C>, Proxy<C::Request, C::Response>) {
+    let (tx, rx) = make_proxy();
+
+    let handler = ClientHandler::<C> {
+        requests: rx,
+        in_flight: VecDeque::with_capacity(32),
+    };
+
+    (handler, tx)
```
Done.

kinetiknz commented on audioipc/src/rpccore.rs:
```diff
@@ -0,0 +1,219 @@
+
+// RPC Client Proxy implementation
+type ProxyRequest<R, Q> = (R, mpsc::Sender<Q>);
+type ProxyReceiver<R, Q> = mpsc::Receiver<ProxyRequest<R, Q>>;
```
Renamed these to `Response` and `Request` everywhere.

kinetiknz commented on audioipc/src/rpccore.rs:
```diff
@@ -0,0 +1,219 @@
+    // Note: In/Out types are swapped compared to ServerHandler, as
+    // clients send requests and receive responses.
+    type In = C::Response;
+    type Out = C::Request;
```
Agree it's confusing.  I've tried renaming the types to ClientMessage and ServerMessage where appropriate; I think it's a bit clearer now.

kinetiknz commented on audioipc/src/rpccore.rs:
```diff
@@ -0,0 +1,219 @@
+}
+
+// RPC Client Proxy implementation
+type ProxyRequest<R, Q> = (R, mpsc::Sender<Q>);
```
Done.

kinetiknz commented on audioipc/src/ipccore.rs:
```diff
@@ -0,0 +1,717 @@
+                self.token,
+                self.outbound
+            );
+        }
```
Yeah, processing inbound messages will result in outbound messages being produced, which we then want to send on the connection.

kinetiknz commented on audioipc/src/rpccore.rs:
```diff
@@ -0,0 +1,219 @@
+
+        // Try to get a new request
+        match self.requests.try_recv() {
+            Ok((request, complete)) => {
```
Done.

kinetiknz commented on audioipc/src/rpccore.rs:
```diff
@@ -0,0 +1,219 @@
+        // Try to get a new request
+        match self.requests.try_recv() {
+            Ok((request, complete)) => {
+                trace!("  --> received request");
```
Let's consider adding this later if it'll be useful for debugging.

kinetiknz commented on audioipc/src/ipccore.rs:
```diff
@@ -0,0 +1,717 @@
+            trace!("connection slab full, insert will allocate");
+        }
+        let entry = self.connections.vacant_entry();
+        let token = Token(entry.key());
```
Good idea, added.

kinetiknz commented on audioipc/src/ipccore.rs:
```diff
@@ -0,0 +1,717 @@
+
+    // Handle readiness event.  Errors returned are fatal for this connection, resulting in removal from the EventLoop connection list.
+    // The EventLoop will call this for any connection that has received an event, including external wake events to clear the outbound buffer.
+    fn handle_event(&mut self, event: Option<&Event>, registry: &Registry) -> Result<bool> {
```
`Connection` doesn't have access to the `poll` on `EventLoop`, so not directly.  It'd be possible to use `Registry::try_clone` to store a registry instance per connection, but I'm not sure if that's an improvement over passing in a reference.

kinetiknz commented on audioipc/src/ipccore.rs:
```diff
@@ -0,0 +1,717 @@
+    // Handle readiness event.  Errors returned are fatal for this connection, resulting in removal from the EventLoop connection list.
+    // The EventLoop will call this for any connection that has received an event, including external wake events to clear the outbound buffer.
+    fn handle_event(&mut self, event: Option<&Event>, registry: &Registry) -> Result<bool> {
+        if let Some(event) = event {
```
Thanks for the suggestion!  I've refactored `handle_event` into smaller functions, hopefully it's more readable now.

kinetiknz commented on server/src/server.rs:
```diff
@@ -550,49 +540,42 @@ impl CubebServer {
+                };
+
+                // XXX: should this be on callback or core?
+                // XXX: we should send use client_pipe and send server_pipe, but into_raw() is tricky for sys::Pipe.
```
Sorry, that was a badly worded comment but you've understood it correctly (`NamedPipe` lacks an `into_raw_handle`).  I've tried to improve the comment.

Note that the old code also passed the "wrong" pipe here for similar reasons, so the behaviour should be unchanged.

kinetiknz commented on audioipc/src/sys/mod.rs:
```diff
@@ -0,0 +1,44 @@
+
+pub trait RecvMsg {
+    // Receive data from the associated connection.  `recv_msg` expects the capacity of
+    // the `ConnectionBuffer` members has been adjusted appropriate by the caller.
```
Oops, fixed.

kinetiknz commented on audioipc/src/ipccore.rs:
```diff
@@ -0,0 +1,717 @@
+            if event.is_readable() {
+                loop {
+                    trace!("[{:?}]: pre-recv inbound: {:?}", self.token, self.inbound);
+                    let r = self.io.recv_msg(&mut self.inbound);
```
`IPC_CLIENT_BUFFER_SIZE` is an arbitrary starting buffer size, it's not expected to be large enough for all possible messages.  When `decode` is called, it will always ensure there's sufficient space available to read the next message length prefix (at message start) or next entire message (once message length is known from the length prefix).  The EventLoop will continue reading and attempting to decode as long as there's data available on the connection, causing `decode` to adjust the buffer size as necessary.

It'd be better if the buffer size management was clearer, I'll think about how it might be improved.

kinetiknz commented on audioipc/src/sys/mod.rs:
```diff
@@ -0,0 +1,44 @@
+    // Receive data from the associated connection.  `recv_msg` expects the capacity of
+    // the `ConnectionBuffer` members has been adjusted appropriate by the caller.
+    fn recv_msg(&mut self, buf: &mut ConnectionBuffer) -> Result<usize>;
+}
```
Thanks, removed.

kinetiknz commented on audioipc/src/sys/mod.rs:
```diff
@@ -0,0 +1,44 @@
+    // Send data on the associated connection.  `send_msg` adjusts the length of the
+    // `ConnectionBuffer` members based on the size of the successful send operation.
+    fn send_msg(&mut self, buf: &mut ConnectionBuffer) -> Result<usize>;
+}
```
Thanks, removed.

kinetiknz commented on audioipc/src/sys/unix/mod.rs:
```diff
@@ -0,0 +1,126 @@
+        let flags = 0;
+        let r = unsafe {
+            let mut iovec = [<&mut IoVec>::from(buf.buf.bytes_mut())];
+            msg::recv_msg_with_flags(self.0.as_raw_fd(), &mut iovec, buf.cmsg.bytes_mut(), flags)
```
Moved msg.rs (and cmsg.rs) into audioipc/src/sys/unix.

kinetiknz commented on audioipc/src/sys/unix/mod.rs:
```diff
@@ -0,0 +1,126 @@
+        assert!(!buf.buf.is_empty());
+        let r = {
+            let iovec = [<&IoVec>::from(&buf.buf[..buf.buf.len()])];
+            msg::send_msg_with_flags(self.0.as_raw_fd(), &iovec, &buf.cmsg[..buf.cmsg.len()], 0)
```
Moved msg.rs (and cmsg.rs) into audioipc/src/sys/unix.

kinetiknz commented on audioipc/src/sys/unix/mod.rs:
```diff
@@ -0,0 +1,126 @@
+            Ok(n) => {
+                buf.buf.advance(n);
+                // Close sent fds.
+                let b = buf.cmsg.clone().freeze();
```
`b.is_empty` is true for any messages without an associated handle being sent.  Since this layer doesn't know anything about messages, it needs to handle both cases.  When `b.is_empty` is true, `cmsg::iterator` returns `None`.

kinetiknz commented on audioipc/src/ipccore.rs:
```diff
@@ -0,0 +1,717 @@
+                    assert_eq!(fd.len(), 1);
+                    assert!(handle.is_none());
+                    handle = Some(fd[0]);
+                }
```
Related to the answer above for `send_msg` - `b.is_empty` is true when the received message has no associated handle.

kinetiknz commented on audioipc/src/rpccore.rs:
```diff
@@ -0,0 +1,219 @@
+// trigger response completion by sending the response via a channel
+// connected to a ProxyResponse.
+pub struct ClientHandler<C: Client> {
+    requests: ProxyReceiver<C::Request, C::Response>,
```
I've renamed this to `messages`, does that seem more suitable?

kinetiknz commented on server/src/server.rs:
```diff
@@ -550,49 +540,42 @@ impl CubebServer {
-                }
+                };
+
+                // XXX: should this be on callback or core?
```
Agreed.  I can't recall why this was using the callback thread in the old code.  I've moved it to the RPC thread now.

kinetiknz commented:
Thanks for the great feedback!  Really appreciate the time and thought you've put in to this.  I've addressed most of your comments now (forgot to re-request review when I pushed the bulk of the changes addressing your comments on Monday. 🤦‍♂️).  I'll keep working on the few that are still outstanding and put them up soon.

I've also included a series of small fixes and improvements as individual changes.  I'll squash those into the main commits before merging.

ChunMinChang commented on audioipc/src/codec.rs:
```diff
@@ -159,29 +162,22 @@ where
 
     fn encode(&mut self, item: Self::In, buf: &mut BytesMut) -> io::Result<()> {
```
Is the performance bottleneck here? Is it related to memory allocations?

kinetiknz replied:
There are speed ups in both `encode` and `decode`.  I noticed these while upgrading to bytes 1.1 (not included, needs more work) where I accidentally made serialization slow.  I didn't profile before/after (speed up numbers are from the included criterion benchmark), so I don't have a precise explanation for where the improvements are.  In `decode` it seems to be mainly from avoiding the split & freeze.  Same with `encode` (using the extra buffer to calculate the length prefix after encoding), plus switching from passing a `Writer` in favour of the extra buffer into `serialize_into`.

ChunMinChang reviewed (approved):
Great improvements! Thanks a lot!

kinetiknz commented on ipctest/src/main.rs:
```diff
@@ -154,16 +162,30 @@ fn run_client() -> Result<()> {
     } else {
+        eprintln!("AudioIPC client (pid {})", std::process::id());
+        if wait_for_debugger {
+            eprintln!("Waiting for debugger to attach; hit enter to continue.");
```
It's reading a whole line, so it won't unblock until it sees a newline.  It's true that it doesn't check the input, so any keys pressed before enter are ignored.
